srml/bridge/src/justification.rs

```diff
@@ -0,0 +1,288 @@
+// Instead we'll use a wrapper which implements `Ord` by leveraging the fact that
+// `Block::Hash` implements `AsRef<u8>`, which itself implements `Ord`
+#[derive(Eq)]
+struct BlockHashKey<Block: BlockT>(Block::Hash);
+
+impl<Block: BlockT> BlockHashKey<Block> {
+	fn new(hash: Block::Hash) -> Self {
+		Self(hash)
+	}
+}
+
+impl<Block: BlockT> Ord for BlockHashKey<Block> {
+	fn cmp(&self, other: &Self) -> Ordering {
+		self.0.as_ref().cmp(other.0.as_ref())
+	}
+}
+
+impl<Block: BlockT> PartialOrd for BlockHashKey<Block> {
+	fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+		Some(self.0.as_ref().cmp(other.0.as_ref()))
+	}
+}
+
+impl<Block: BlockT> PartialEq for BlockHashKey<Block> {
+	fn eq(&self, other: &Self) -> bool {
+		self.0.as_ref() == other.0.as_ref()
+	}
+}
+
+/// A utility trait implementing `grandpa::Chain` using a given set of headers.
+/// This is useful when validating commits, using the given set of headers to
+/// verify a valid ancestry route to the target commit block.
+struct AncestryChain<Block: BlockT> {
+	ancestry: BTreeMap<BlockHashKey<Block>, Block::Header>,
+}
+
+impl<Block: BlockT> AncestryChain<Block> {
+	fn new(ancestry: &[Block::Header]) -> AncestryChain<Block> {
+		let ancestry: BTreeMap<_, _> = ancestry
+			.iter()
+			.cloned()
+			.map(|h: Block::Header| (BlockHashKey::new(h.hash()), h))
+			.collect();
+
+		AncestryChain { ancestry }
+	}
+}
+
+impl<Block: BlockT> grandpa::Chain<Block::Hash, NumberFor<Block>> for AncestryChain<Block> where
+	NumberFor<Block>: grandpa::BlockNumberOps
+{
+	fn ancestry(&self, base: Block::Hash, block: Block::Hash) -> Result<Vec<Block::Hash>, GrandpaError> {
+		let mut route = Vec::new();
+		let mut current_hash = block;
+		loop {
+			if current_hash == base { break; }
+
+			let key = BlockHashKey::new(current_hash);
+			match self.ancestry.get(&key) {
+				Some(current_header) => {
+					current_hash = *current_header.parent_hash();
+					route.push(current_hash);
+				},
+				_ => return Err(GrandpaError::NotDescendent),
+			}
+		}
+		route.pop(); // remove the base
+
+		Ok(route)
+	}
+
+	fn best_chain_containing(&self, _block: Block::Hash) -> Option<(Block::Hash, NumberFor<Block>)> {
+		None
+	}
+}
+
+pub(crate) fn localized_payload<E: Encode>(round: RoundNumber, set_id: SetIdNumber, message: &E) -> Vec<u8> {
+	(message, round, set_id).encode()
+}
+
+// Check the signature of a Grandpa message.
+// This was originally taken from `communication/mod.rs`
+fn check_message_sig<Block: BlockT>(
+	message: &Message<Block>,
+	id: &AuthorityId,
+	signature: &AuthoritySignature,
+	round: RoundNumber,
+	set_id: SetIdNumber,
+) -> Result<(), ()> {
+	let as_public = id.clone();
+	let encoded_raw = localized_payload(round, set_id, message);
+
+	// Since `app::Public` implements `RuntimeAppPublic` we can call `verify()`
+	if as_public.verify(&encoded_raw, signature) {
+		Ok(())
+	} else {
+		// debug!(target: "afg", "Bad signature on message from {:?}", id);
```

Comment:
We probably want to keep this for the client code that uses it once this is refactored to re-use the same code.